### ISSUE TITLE
Use expression for tags instead of include/exclude

### ISSUE THREAD

**Issue by alladywek** [kotest/kotest#863]
Tags are defenitely good idea to group your tests but current include/exclude implementation doesn't provide enough flexibility.
I want to use simple expressions like in [Junit5](https://junit.org/junit5/docs/current/user-guide/#running-tests-tag-expressions)
It will give users write simple and readable expressions:

> kotlintest.tags.exp="tag1 & tag2"
> kotlintest.tags.exp="smoke & (tag1 | tag2)"
> kotlintest.tags.exp="smoke & !tag2"

It could be implemented as extension or replace current solution in next release.

What do you think?

**Comment by LeoColman** [comment]
I'm not sure if I understood why `include/exclude` is inflexibile, can you develop a little bit more on that idea?

**Comment by sksamuel** [comment]
I guess it's like run this test only if tag x is present and tag y is not present

**Comment by alladywek** [comment]
```
class A : StringSpec() {
    override fun tags(): Set<Tag> = setOf(Spec1)

    init {
        "Atest1".config(tags = setOf(Tag1)) { println("Atest1") }
        "Atest2".config(tags = setOf(Tag2)) { println("Atest2") }
        "Atest3".config(tags = setOf(Tag1, Tag2)) { println("Atest3") }
    }
}

class B : StringSpec() {
    override fun tags(): Set<Tag> = setOf(Spec2)

    init {
        "Btest1".config(tags = setOf(Tag1)) { println("Btest1") }
        "Btest2".config(tags = setOf(Tag2)) { println("Btest2") }
    }
}
```
How can we run only `Atest3`? 
with expression:
>`Spec1 & Tag1 & Tag2` 

or if only spec `A` exists:
>`Tag1 & Tag2`

How can we run only `Atest1` and `Atest2`?
with expression:
>`Spec1 & (Tag1 | Tag2)`

or if only spec `A` exists:
>`Tag1 | Tag2`

What if I need run tests with `Tag2` only from spec `A` and `B`, but I have a lot of other specs where I use `Tag2`?
My point that expression is quite easy way to configure it with one property and readable syntax.

**Comment by sksamuel** [comment]
I guess the workaround for now is to create another tag, Tag3 which is used when you have Tag1 and Tag2.

Adding the expression syntax is a good idea, and I can think we can do something for 3.5

**Comment by sksamuel** [comment]
This is now included in 4.1.2
